src/app/app-interactive-map/app-interactive-map.component.ts

```diff
@@ -127,23 +133,79 @@ export class AppInteractiveMapComponent implements OnInit, AfterViewInit {
+      direction: this.card.knockoutReactions.includes(reactionId) ? OperationDirection.Undo : OperationDirection.Do,
+    }));
   }
 
-  handleSetAsObjective(args: string): void {
-    console.log('SET AS OBJECTIVE', args);
+  handleSetAsObjective(reactionId: string): void {
+    this.store.dispatch(new fromActions.SetObjectiveReaction({
+      reactionId,
+      direction: 'max',
+    }));
   }
 
-  handleChangeBounds(args: string): void {
-    console.log('CHANGE BOUNDS', args);
+  handleChangeBounds(reactionId: string): void {
+    this.store.dispatch(new fromActions.ReactionOperation({
+      item: {
+        reaction: this.findReaction(this.card.model.reactions, reactionId),
+        lowerBound: parseInt(this.elRef.nativeElement.querySelector('#lowerbound').value, 10),
+        upperBound: parseInt(this.elRef.nativeElement.querySelector('#upperbound').value, 10),
+      },
+      operationTarget: 'bounds',
+      direction: OperationDirection.Do,
+    }));
   }
 
-  handleResetBounds(args: string): void {
-    console.log('RESET BOUNDS', args);
+  handleResetBounds(reactionId: string): void {
+    this.store.dispatch(new fromActions.ReactionOperation({
+      item: <Bound>(this.findBoundReaction(reactionId)),
+      operationTarget: 'bounds',
+      direction: OperationDirection.Undo,
+    }));
   }
 
-  handleObjectiveDirection(args: string): void {
-    console.log('OBJECTIVE DIRECTION', args);
+  handleObjectiveDirection(reactionId: string): void {
+    this.store.dispatch(new fromActions.SetObjectiveReaction({
+      reactionId,
+      direction: this.card.objectiveReaction.direction === 'max' ? 'min' : 'max',
+    }));
+  }
+
+  checkStateObjective(reactionId: string): boolean {
+    if (this.card.objectiveReaction) {
+      return (this.card.objectiveReaction.reactionId === reactionId && this.card.objectiveReaction.direction === 'max');
+    } else {
+      return false;
+    }
+  }
+
+  upperBound(reactionId: string): number {
+    const reaction = this.findBoundReaction(reactionId);
+    return (<Cobra.Reaction>reaction).upper_bound ? (<Cobra.Reaction>reaction).upper_bound : (<Bound>reaction).upperBound;
+  }
+
+  lowerBound(reactionId: string): number {
+    const reaction = this.findBoundReaction(reactionId);
+    return (<Cobra.Reaction>reaction).lower_bound ? (<Cobra.Reaction>reaction).upper_bound : (<Bound>reaction).lowerBound;
+  }
+
+  knockoutText(reactionId: string): string {
+    return this.card.knockoutReactions && this.card.knockoutReactions.includes(reactionId) ? 'Undo knockout' : 'Knockout';
+  }
+
+  objectiveText(reactionId: string): string {
+    return this.card.objectiveReaction && this.card.objectiveReaction.reactionId === reactionId ? 'Undo set as objective' : 'Set as objective';
+  }
+
+  findBoundReaction (reactionId: string): Bound | Cobra.Reaction {
+    return this.card.bounds.find((r) => r.reaction.id === reactionId) || this.findReaction(this.card.model.reactions, reactionId);
```

Comment:
Why can this return both `Bound` and `Cobra.Reaction`? I think with the duality, having to cast types and handle different casing (like `upper_bound` vs `upperBound`) is pretty ugly and shouldn't be necessary.

The serialized model will not be stored in the frontend state when the model service is able to fetch models itself from the model warehouse, so `card.model` will just be a model id reference.

So why is it uncertain whether the reaction can be found in `this.cards.bounds` (a list of changed bounds, from what I understand) or in the model (`this.card.model`)? I don't see in what cases it is necessary to get the reaction from the model - but we need to deal with not having it available.

Reply:
The problem here is, I have to access to the model, to know the first bounds, when you dont have anything inside of the this.cards.bound, also, I am not sure about the structure of do everything inside the same function.